Unable to run application with kafka-stream on aarch64 built using native builder image

### Describe the bug

Unable to run application with kafka-stream on aarch64 in native, application built using builder image (use-case of RHBQ)

application startup fails with
```
[hudson@ampere-mtsnow-altramax-03-vm-13 code-with-quarkus]$ target/code-with-quarkus-1.0.0-SNAPSHOT-runner
Sep 15, 2024 7:18:35 AM io.quarkus.runtime.ApplicationLifecycleManager run
ERROR: Failed to start application
java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:119)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:124)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
Caused by: java.lang.ExceptionInInitializerError
	at io.quarkus.kafka.streams.runtime.KafkaStreamsRecorder.loadRocksDb(KafkaStreamsRecorder.java:14)
	at io.quarkus.deployment.steps.KafkaStreamsProcessor$loadRocksDb1611413226.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.KafkaStreamsProcessor$loadRocksDb1611413226.deploy(Unknown Source)
	... 7 more
Caused by: java.lang.RuntimeException: librocksdbjni-linux-aarch64.so was not found inside JAR.
	at org.rocksdb.NativeLibraryLoader.loadLibraryFromJarToTemp(NativeLibraryLoader.java:118)
	at org.rocksdb.NativeLibraryLoader.loadLibraryFromJar(NativeLibraryLoader.java:102)
	at org.rocksdb.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:82)
	at org.rocksdb.RocksDB.loadLibrary(RocksDB.java:70)
	at org.rocksdb.RocksDB.<clinit>(RocksDB.java:39)
	... 10 more
```

The binary contains strings with mixture of `librocksdbjni-linux-aarch64.so` and `librocksdbjni-linux64.so`
```
strings target/code-with-quarkus-1.0.0-SNAPSHOT-runner | grep librocksdbjni
librocksdbjni-linux-aarch64.so
librocksdbjni-linux64.so
librocksdbjni
```

If the application is built using locally installed Mandrel, the app works well and strings command returns just librocksdbjni-linux-aarch64.so in 2 occurrences.

`docker inspect quay.io/quarkus/ubi-quarkus-mandrel-builder-image:jdk-21` reports architecture as `arm64`

This was detected when working on investigation of Quarkus Quickstarts failures investigation on aarch64 platform.

CC @cescoffier @zakkak @galderz
FYI @mjurc

### Expected behavior

Native application built using native builder image can start on RHEL 8 aarch64

### Actual behavior

Native application built using native builder image can't start on RHEL 8 aarch64

### How to Reproduce?

Using code.quarkus:
```
wget -O app.zip "https://code.quarkus.io/d?e=kafka-streams&cn=code.quarkus.io"
unzip -q app.zip
cd code-with-quarkus/
mvn clean package -Dnative -Dquarkus.native.container-build=true
target/code-with-quarkus-1.0.0-SNAPSHOT-runner
```

Using https://github.com/quarkusio/quarkus-quickstarts:
```
git clone https://github.com/quarkusio/quarkus-quickstarts -b development
cd quarkus-quickstarts
mvn clean verify -f kafka-streams-quickstart -Dnative -Dquarkus.native.container-build=true
```

### Output of `uname -a` or `ver`

RHEL 8 aarch64

### Output of `java -version`

Java 21

### Mandrel or GraalVM version (if different from Java)

quay.io/quarkus/ubi-quarkus-mandrel-builder-image:jdk-21

### Quarkus version or git rev

Quarkus main / Quarkus 3.14

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/extensions/kafka-streams/deployment/src/main/java/io/quarkus/kafka/streams/deployment/KafkaStreamsProcessor.java
+++ b/extensions/kafka-streams/deployment/src/main/java/io/quarkus/kafka/streams/deployment/KafkaStreamsProcessor.java
@@ -2,7 +2,6 @@ package io.quarkus.kafka.streams.deployment;
 
 import static io.quarkus.kafka.streams.runtime.KafkaStreamsPropertiesUtil.buildKafkaStreamsProperties;
 
-import java.io.IOException;
 import java.util.Properties;
 
 import jakarta.inject.Singleton;
@@ -16,7 +15,6 @@ import org.apache.kafka.streams.processor.internals.DefaultKafkaClientSupplier;
 import org.apache.kafka.streams.processor.internals.StreamsPartitionAssignor;
 import org.rocksdb.RocksDBException;
 import org.rocksdb.Status;
-import org.rocksdb.util.Environment;
 
 import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
 import io.quarkus.arc.deployment.SyntheticBeanBuildItem;
@@ -28,16 +26,16 @@ import io.quarkus.deployment.annotations.ExecutionTime;
 import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.deployment.builditem.LaunchModeBuildItem;
+import io.quarkus.deployment.builditem.NativeImageFeatureBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.JniRuntimeAccessBuildItem;
-import io.quarkus.deployment.builditem.nativeimage.NativeImageResourceBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.RuntimeReinitializedClassBuildItem;
-import io.quarkus.deployment.pkg.builditem.NativeImageRunnerBuildItem;
 import io.quarkus.deployment.pkg.steps.NativeOrNativeSourcesBuild;
 import io.quarkus.kafka.streams.runtime.KafkaStreamsProducer;
 import io.quarkus.kafka.streams.runtime.KafkaStreamsRecorder;
 import io.quarkus.kafka.streams.runtime.KafkaStreamsRuntimeConfig;
 import io.quarkus.kafka.streams.runtime.KafkaStreamsSupport;
+import io.quarkus.kafka.streams.runtime.graal.KafkaStreamsFeature;
 import io.quarkus.smallrye.health.deployment.spi.HealthBuildItem;
 
 class KafkaStreamsProcessor {
@@ -53,12 +51,9 @@ class KafkaStreamsProcessor {
     void build(BuildProducer<ReflectiveClassBuildItem> reflectiveClasses,
             BuildProducer<JniRuntimeAccessBuildItem> jniRuntimeAccessibleClasses,
             BuildProducer<RuntimeReinitializedClassBuildItem> reinitialized,
-            BuildProducer<NativeImageResourceBuildItem> nativeLibs,
-            LaunchModeBuildItem launchMode,
-            NativeImageRunnerBuildItem nativeImageRunner) throws IOException {
+            LaunchModeBuildItem launchMode) {
         registerClassesThatAreLoadedThroughReflection(reflectiveClasses, launchMode);
         registerClassesThatAreAccessedViaJni(jniRuntimeAccessibleClasses);
-        addSupportForRocksDbLib(nativeLibs, nativeImageRunner);
         enableLoadOfNativeLibs(reinitialized);
     }
 
@@ -174,18 +169,6 @@ class KafkaStreamsProcessor {
                 .produce(new JniRuntimeAccessBuildItem(true, false, false, RocksDBException.class, Status.class));
     }
 
-    private void addSupportForRocksDbLib(BuildProducer<NativeImageResourceBuildItem> nativeLibs,
-            NativeImageRunnerBuildItem nativeImageRunnerFactory) {
-        // for RocksDB, either add linux64 native lib when targeting containers
-        if (nativeImageRunnerFactory.isContainerBuild()) {
-            nativeLibs.produce(new NativeImageResourceBuildItem("librocksdbjni-linux64.so"));
-        }
-        // otherwise the native lib of the platform this build runs on
-        else {
-            nativeLibs.produce(new NativeImageResourceBuildItem(Environment.getJniLibraryFileName("rocksdb")));
-        }
-    }
-
     private void enableLoadOfNativeLibs(BuildProducer<RuntimeReinitializedClassBuildItem> reinitialized) {
         reinitialized.produce(new RuntimeReinitializedClassBuildItem("org.rocksdb.RocksDB"));
     }
@@ -235,5 +218,10 @@ class KafkaStreamsProcessor {
                 new HealthBuildItem(
                         "io.quarkus.kafka.streams.runtime.health.KafkaStreamsStateHealthCheck",
                         buildTimeConfig.healthEnabled));
+    }
+
+    @BuildStep
+    NativeImageFeatureBuildItem kafkaStreamsFeature() {
+        return new NativeImageFeatureBuildItem(KafkaStreamsFeature.class.getName());
     }
 }

--- a/extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/graal/KafkaStreamsFeature.java
+++ b/extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/graal/KafkaStreamsFeature.java
@@ -1,0 +1,15 @@
+package io.quarkus.kafka.streams.runtime.graal;
+
+import org.graalvm.nativeimage.hosted.Feature;
+import org.graalvm.nativeimage.hosted.RuntimeResourceAccess;
+import org.rocksdb.util.Environment;
+
+public class KafkaStreamsFeature implements Feature {
+
+    @Override
+    public void afterRegistration(AfterRegistrationAccess access) {
+        String libraryFileName = Environment.getJniLibraryFileName("rocksdb");
+        RuntimeResourceAccess.addResource(Environment.class.getModule(), libraryFileName);
+    }
+
+}

